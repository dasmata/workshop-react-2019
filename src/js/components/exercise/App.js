/**
 * Components exercise
 *
 * Convert all class components in function components
 *
 * */

import React, { Fragment, Component } from 'react';
import HelloWorld from './HelloWorld';

class App extends Component {

    handleMenuClick = (e) => {
        e.preventDefault();
        console.log(e.target.title);
    };

    render () {
        return (
            <Fragment>
                <section>
                    <header>
                        <ul>
                            <li><a href={'javascript://'} onClick={this.handleMenuClick} title={'first item'}>first item</a></li>
                            <li><a href={'javascript://'} onClick={this.handleMenuClick} title={'second item'}>second item</a></li>
                            <li><a href={'javascript://'} onClick={this.handleMenuClick} title={'third item'}>third item</a></li>
                        </ul>
                    </header>
                </section>
                <section>
                    <main>
                        <HelloWorld/>
                    </main>
                </section>
                <section>
                    <footer></footer>
                </section>
            </Fragment>
        );
    }
}

export default App;