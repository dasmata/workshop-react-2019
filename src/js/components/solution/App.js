import React, { Fragment, useEffect, useState } from 'react';
import HelloWorld from './HelloWorld';
import AppMenu from './AppMenu';
import history from './browserHistory';

const PAGE_HOME = "/";
const PAGE_SECOND = "/menu-2";
const PAGE_THIRD = "/menu-3";

const menuConfig = [
    {
        text: 'menu 1',
        url: PAGE_HOME,
        title: "Homepage"
    },
    {
        text: 'menu 2',
        url: PAGE_SECOND,
        title: "second page"
    },
    {
        text: 'menu 3',
        url: PAGE_THIRD,
        title: "third page"
    },
];

const PAGES = {
    [PAGE_HOME]: HelloWorld,
    [PAGE_SECOND]: ()=>{return <div>second page</div>},
    [PAGE_THIRD]: ()=>{return <div>third page</div>}
};

const App = (props) => {
    const [pageURL, setPageURL] = useState(location.pathname);
    let Component = PAGES[pageURL];


    useEffect(()=>{
        history.listen((location, action) => {
            setPageURL(location.state.url);
            document.title = location.state.title;
        });
        document.title = menuConfig.reduce((title, el)=>{
            if(el.url === pageURL){
                return el.title;
            }
            return title;
        }, '');
    }, [1]);

    return (
        <Fragment>
            <section>
                <header>
                    <AppMenu config={menuConfig}/>
                </header>
            </section>
            <section>
                <main>
                    <Component/>
                </main>
            </section>
            <section>
                <footer></footer>
            </section>
        </Fragment>
    );
};

export default App;