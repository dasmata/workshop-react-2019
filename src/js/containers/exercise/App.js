/**
 * CONTAINERS EXERCISE
 *
 * Refactor the App component in such a way that all the logic
 * is moved in a LoginContainer.
 * LoginContainer should implement a on login method to notify App Componennt
 * when the login occurs so that App can display de welcome message
 */
import React from 'react';
import Loader from './Loader/index';

const submitLogin = (username, password) => {
    return new Promise((done, fail) => {
        setTimeout(() => {
            done();
        }, 2000);
    });
};

export default class App extends React.Component {

    state = {
        username: '',
        password: '',
        loading: false,
        auth: false
    };

    submit () {
        if (this.state.username === '' || this.state.password === '') {
            console.error('Invalid form data');
            return;
        }
        this.setState({loading: true}, () => {
            submitLogin(this.state.username, this.state.password).then(() => {
                this.setState({
                    loading: false,
                    auth: true
                });
            });
        });

    }

    render () {
        return (<Loader loading={this.state.loading}>
            {!this.state.auth && <form onSubmit={(e) => {
                e.preventDefault();
                this.submit();
            }}>
                <label>
                    <div>Username</div>
                    <input
                        name={'username'}
                        placeholder={'ex:johndoe2014'}
                        onChange={(e) => this.setState({username: e.target.value})}
                        value={this.state.username}
                    />
                </label>
                <label>
                    <div>Password</div>
                    <input
                        name={'password'}
                        type={'password'}
                        placeholder={'your password'}
                        onChange={(e) => this.setState({password: e.target.value})}
                        value={this.state.password}
                    />
                </label>
                <div>
                    <button>Login</button>
                </div>
            </form>}
            {this.state.auth && <h1>WELCOME {this.state.username}!!!</h1>}
        </Loader>);
    }
}