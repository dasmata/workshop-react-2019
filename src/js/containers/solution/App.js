import React from 'react';
import LoginContainer from './LoginContainer';
import WelcomeComponent from './WelcomeComponent';

class App extends React.Component {

    state = {
        user: null,
    };

    onLogin = (user) => {
        this.setState({user});
    };

    render () {
        return this.state.user === null ? (<LoginContainer onLogin={this.onLogin}/>) : (<WelcomeComponent username={this.state.user.username}/>);
    }
}

export default App;