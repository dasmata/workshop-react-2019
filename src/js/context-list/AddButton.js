import React, {useContext} from 'react';
import ListContext from './listContext';

export default (props) => {
    const ctx = useContext(ListContext);

    return (<a href={'javascript://'} onClick={ctx.onShowForm}>Add element</a>)
}