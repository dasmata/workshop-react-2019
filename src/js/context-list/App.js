import React from 'react';
import List from './List';

export default () => {
    return (
        <List>
            <List.Form />
            <List.Grid />
            <List.AddButton />
        </List>
    )
}