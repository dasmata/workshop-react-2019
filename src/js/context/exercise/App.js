/**
 * Implement the basic CR(U)D functionality by using react context
 * Use App's state as a context containing with all the items and the callbacks
 * Use App to render the context provider and to provide the context value:
 * <Context.Provider value>
 *   <ElementsList/>
 *   <AddButton/>
 *   <Form/>
 * </Context.Provider>
 * ElementsList, AddButton and Form should all be context consumers that either
 * render elements or call callbacks that change elements list.
 *
 */
import React, { Component, Fragment } from 'react';
import ElementsList from './ElementsList';
import Form from './Form';

export default class App extends Component {

    state = {
        items: [],
        showForm: false,
    };

    deleteElement = (element) => {
        this.setState({
            items: this.state.items.filter((el) => {return el !== element;}),
        });
    };

    showForm = () => {
        this.setState({
            showForm: true,
        });
    };

    onAdd = (text) => {
        this.setState({
            items: [...this.state.items, {id: this.state.items.length + 1, text}],
            showForm: false,
        });
    };

    render () {
        return (
            <Fragment>
                <ElementsList items={this.state.items} onDelete={this.deleteElement}/>
                <a href={'javascript://'} onClick={this.showForm}>+ Add element</a>
                <Form show={this.state.showForm} onAdd={this.onAdd}/>
            </Fragment>

        );
    }

}