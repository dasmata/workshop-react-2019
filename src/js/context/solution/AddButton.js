import React from 'react';
import Context from './Context';

export default () => (
    <Context.Consumer>
        {(context)=>(
            <a href={'javascript://'} onClick={context.onShowForm}>+ Add element</a>
        )}
    </Context.Consumer>
);