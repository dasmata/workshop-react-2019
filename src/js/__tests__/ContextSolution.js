import React from 'react';
import { configure, shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

import App from '../context/solution/App';
import AddButton from '../context/solution/AddButton';

require('jsdom-global')();
configure({adapter: new Adapter()});

describe('Testing the context solution', () => {

    it('renders initial snapshop', () => {
        expect.assertions(1);
        const tree = render(
            <App/>,
        );
        expect(toJson(tree)).toMatchSnapshot();
    });

    it('add element', () => {
        expect.assertions(2);
        const tree = mount(
            <App/>,
        );
        tree.find('a').at(0).simulate('click');
        expect(tree.find(App).state('showForm')).toBe(true);
        expect(toJson(tree)).toMatchSnapshot();
    });
});
