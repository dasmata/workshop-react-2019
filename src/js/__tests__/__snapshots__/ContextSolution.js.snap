// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the context solution add element 1`] = `
<App>
  <_default>
    <ul />
  </_default>
  <_default>
    <a
      href="javascript://"
      onClick={[Function]}
    >
      + Add element
    </a>
  </_default>
  <Form>
    <div
      style={
        Object {
          "display": "block",
        }
      }
    >
      <input
        onChange={[Function]}
        placeholder="item text"
        type="text"
        value=""
      />
      <button
        onClick={[Function]}
      >
        Add
      </button>
    </div>
  </Form>
</App>
`;

exports[`Testing the context solution renders initial snapshop 1`] = `
Array [
  <ul />,
  <a
    href="javascript://"
  >
    + Add element
  </a>,
  <div
    style="display:none"
  >
    <input
      placeholder="item text"
      type="text"
      value=""
    />
    <button>
      Add
    </button>
  </div>,
]
`;
